src/Orleans.Serialization/Codecs/StackCodec.cs

```diff
@@ -7,165 +7,166 @@
+        writer.WriteFieldHeader(fieldIdDelta, expectedType, value.GetType(), WireType.TagDelimited);
+
+        if (value.Count > 0)
         {
-            if (ReferenceCodec.TryWriteReferenceField(ref writer, fieldIdDelta, expectedType, value))
+            UInt32Codec.WriteField(ref writer, 0, (uint)value.Count);
+            uint innerFieldIdDelta = 1;
+            foreach (var element in value)
             {
-                return;
+                _fieldCodec.WriteField(ref writer, innerFieldIdDelta, CodecElementType, element);
+                innerFieldIdDelta = 0;
             }
+        }
 
-            writer.WriteFieldHeader(fieldIdDelta, expectedType, value.GetType(), WireType.TagDelimited);
-
-            if (value.Count > 0)
-            {
-                UInt32Codec.WriteField(ref writer, 0, (uint)value.Count);
-                uint innerFieldIdDelta = 1;
-                foreach (var element in value)
-                {
-                    _fieldCodec.WriteField(ref writer, innerFieldIdDelta, CodecElementType, element);
-                    innerFieldIdDelta = 0;
-                }
-            }
+        writer.WriteEndObject();
+    }
 
-            writer.WriteEndObject();
+    /// <inheritdoc/>
+    public Stack<T> ReadValue<TInput>(ref Reader<TInput> reader, Field field)
+    {
+        if (field.WireType == WireType.Reference)
+        {
+            return ReferenceCodec.ReadReference<Stack<T>, TInput>(ref reader, field);
         }
 
-        /// <inheritdoc/>
-        public Stack<T> ReadValue<TInput>(ref Reader<TInput> reader, Field field)
+        field.EnsureWireTypeTagDelimited();
+
+        var placeholderReferenceId = ReferenceCodec.CreateRecordPlaceholder(reader.Session);
+        T[] array = null;
+        var i = 0;
+        uint fieldId = 0;
+        while (true)
         {
-            if (field.WireType == WireType.Reference)
+            var header = reader.ReadFieldHeader();
+            if (header.IsEndBaseOrEndObject)
             {
-                return ReferenceCodec.ReadReference<Stack<T>, TInput>(ref reader, field);
+                break;
             }
 
-            field.EnsureWireTypeTagDelimited();
-
-            var placeholderReferenceId = ReferenceCodec.CreateRecordPlaceholder(reader.Session);
-            Stack<T> result = null;
-            uint fieldId = 0;
-            while (true)
+            fieldId += header.FieldIdDelta;
+            switch (fieldId)
             {
-                var header = reader.ReadFieldHeader();
-                if (header.IsEndBaseOrEndObject)
-                {
+                case 0:
+                    var length = (int)UInt32Codec.ReadValue(ref reader, header);
+                    if (length > 10240 && length > reader.Length)
+                    {
+                        ThrowInvalidSizeException(length);
+                    }
+
+                    array = new T[length];
+                    i = length - 1;
```

Comment:
The crux of the fix is to populate an array in reverse order and then initialize the stack with it. Stack does not offer a way to get/modify the underlying array directly or enumerate in reverse order.